### PR TITLE
storcon: introduce step down primitive

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem
We are missing the step-down primitive required to implement rolling restarts of the storage controller.

## Summary of changes
Add `/control/v1/step_down` endpoint which puts the storage controller into a state where it rejects
all API requests apart from `/control/v1/step_down`, `/status` and `/metrics`. When receiving the request,
storage controller cancels all pending reconciles and waits for them to exit gracefully. The response contains
a snapshot of the in-memory observed state.

Related:
* https://github.com/neondatabase/cloud/issues/14701
* https://github.com/neondatabase/neon/issues/7797
* https://github.com/neondatabase/neon/pull/8310

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: storage_controller/src/http.rs [hunk 1]
```diff
@@ -734,6 +741,41 @@ struct RequestMeta {
     at: Instant,
 }
 
+pub fn prologue_ledearship_status_check_middleware<
```
Comment (yliang412) [inline storage_controller/src/http.rs:744]:
```suggestion
pub fn prologue_leadership_status_check_middleware<
```

File: storage_controller/src/http.rs [hunk 2]
```diff
@@ -820,6 +862,7 @@ pub fn make_router(
     build_info: BuildInfo,
 ) -> RouterBuilder<hyper::Body, ApiError> {
     let mut router = endpoint::make_router()
+        .middleware(prologue_ledearship_status_check_middleware())
```
Comment (yliang412) [inline storage_controller/src/http.rs:865]:
```suggestion
        .middleware(prologue_leadership_status_check_middleware())
```

File: test_runner/regress/test_storage_controller.py [hunk 3]
```diff
@@ -1783,3 +1783,78 @@ def assert_victim_evacuated():
...
+    assert tenant_conf["compaction_threshold"] == 5
+
+    # Validate that we propagated the change to the pageserver
+    ps_tenant_conf = env.pageserver.http_client().tenant_config(tid)
+    assert "compaction_threshold" in ps_tenant_conf.effective_config
+    assert ps_tenant_conf.effective_config["compaction_threshold"] == 5
+
+    # Validate that the storcon is not replying to the usual requests
+    # once it has stepped down.
+    with pytest.raises(StorageControllerApiException):
```
Comment (yliang412) [inline test_runner/regress/test_storage_controller.py:1833]: could add a `match="Resource temporarily unavailable.*"` parameter while raising the exception to make the 503 checks more explicit.

File: storage_controller/src/service.rs [hunk 4]
```diff
@@ -403,11 +461,30 @@ struct ShardUpdate {
...
+        write!(writer, "{}", s)
+    }
+}
+
 pub(crate) enum ReconcileResultRequest {
     ReconcileResult(ReconcileResult),
     Stop,
 }
 
+// TODO: move this into the storcon peer client when that gets added
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply storage_controller/src/service.rs:484]: That's how a candidate storage controller will ask the leader to step down.

File: storage_controller/src/http.rs [hunk 5]
```diff
@@ -734,6 +741,41 @@ struct RequestMeta {
...
+        let allowed_routes = match leadership_status {
+            LeadershipStatus::Leader => Vec::default(),
+            LeadershipStatus::SteppedDown => {
+                // TODO: does it make sense to allow /status here?
+                ["/control/v1/step_down", "/status", "/metrics"].to_vec()
+            }
+            LeadershipStatus::Candidate => ["/ready", "/status", "/metrics"].to_vec(),
+        };
+
+        if allowed_routes.is_empty() {
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply storage_controller/src/http.rs]: Good point. Added an enum to make things clearer: [0c5193c](https://github.com/neondatabase/neon/pull/8512/commits/0c5193ca8167e7a130121c9a035823e6f8fa5739)

File: storage_controller/src/http.rs (hunk 1 above)
Comment (VladLazar) [reply storage_controller/src/http.rs]: https://github.com/neondatabase/neon/pull/8512/commits/df74f8d5f8522385249bfafc083b553c514448ff

File: storage_controller/src/http.rs (hunk 2 above)
Comment (VladLazar) [reply storage_controller/src/http.rs]: https://github.com/neondatabase/neon/pull/8512/commits/df74f8d5f8522385249bfafc083b553c514448ff

File: test_runner/regress/test_storage_controller.py (hunk 3 above)
Comment (VladLazar) [reply test_runner/regress/test_storage_controller.py]: https://github.com/neondatabase/neon/pull/8512/commits/5205aae2cccfe3a2cf95ec9a63f6f7478178ce85

File: storage_controller/src/service.rs [hunk 6]
```diff
@@ -131,6 +133,15 @@ enum NodeOperations {
     Delete,
 }
 
+#[derive(Copy, Clone, strum_macros::Display, measured::FixedCardinalityLabel)]
+#[strum(serialize_all = "snake_case")]
+pub(crate) enum LeadershipStatus {
+    Leader,
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply storage_controller/src/service.rs:145]: sure https://github.com/neondatabase/neon/pull/8512/commits/14e238a38cbd27f6bc961c3b656a3944e768796a

**Pull request merged**